src/main/java/edu/uwaterloo/ece658/service/DataHandlerService.java

```diff
@@ -0,0 +1,160 @@
+        Tag tag = new Tag();
+        tag.setName(tagName);
+        tagFacade.create(tag);
+        return tag;
+    }
+    
+    private Tag initializeTag(String tagName) {
+        Tag tag = tagFacade.getNormalTagByName(tagName);
+        if (tag == null) {
+            tag = createNewTag(tagName);
+        }
+        return tag;
+    }
+    
+    private List<Tag> initializeTags(List<String> tagNames) {
+        List<Tag> tags = new ArrayList<>();
+        
+        for (String tagName : tagNames) {
+            tags.add(initializeTag(tagName));
+        }
+        return tags;
+    }
+    
+    public boolean updateDatabaseWithPhotoInformation(
+            String username, 
+            String S3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate,
+            boolean isPrivatePhoto,
+            List<String> tagNames)
+    {
+        // store key in table with tags
+        List<Tag> tags = initializeTags(tagNames);
+        User user = userFacade.findUserByUserName(username);
+        Photo photo = initializePhoto(
+                        S3key,
+                        Md5,
+                        photoSize,
+                        uploadDate,
+                        isPrivatePhoto);
+        
+        photo.setTags(tags);
+        photo.addUser(user);
+        
+        user.addUploadedPhoto(photo);
+        userFacade.edit(user);
+        
+        for (Tag tag : tags) {
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);
+        }
+        
+        // call notify to send messages to users
+        return subscriptionService.notifyUsers(tags, photo);
+    }
+    
+    public void removeTagFromPhoto(String username, String S3key, String tagName) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        Tag tag = tagFacade.getNormalTagByName(tagName);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            photo.getTags().remove(tag);
+            tag.getPhotosUnderThisTag().remove(photo);
+            tagFacade.edit(tag);
+            photoFacade.edit(photo);
+        }
+    }
+    
+    public void addTagToPhoto(String username, String S3key, String tagName) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        Tag tag = initializeTag(tagName);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            photo.addTag(tag);
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);
+            photoFacade.edit(photo);
+        }
+    }
+    
+    public boolean deleteImage(String username, String S3key) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        if (photo != null && photo.getUploadedUsers().contains(user)) {
+            user.getUploadedPhotos().remove(photo);

```

Comment:
Like I said before, even if the uploader of a photo should even not be able to deleted the photo because this photo might have been subscribed by other users. We don't hope for the deletion to be propagated to other subscribed users. If the photo is private, or nobody subscribes to the photo by any tag, or has no tags, then it's safe the do the real deletion - deleting from metadata database and the original file from S3. Otherwise, just delink the user and the photo.

Reply:
I think a user should be able to delete their own photo whether its private or not. If they have a unique tag attached that someone is subscribed to either it should be deleted as well or when searching that tag the picture no longer comes up.

Reply:
From technical point we can do that, but I don't think this is good for other users. Your subscription photos might change depending on the uploader deleted the photos or not. As I normal user, I don't hope my subscribed favourite photo gets deleted just because the uploaded deleted it, which harms user experience.